### DISCUSSION TITLE
What are the responsibilities of the MainWindowViewModel?

### DISCUSSION THREAD

**Discussion by coinsoundsbetter** [AvaloniaUI/Avalonia#14154] [Q&A] [answered]
What is the role of the MainWindowViewModel in medium to large projects? Will a lot of code be written to this class?
By reading the official example, I can see that the business logic of other Viewmodels passes through MainWindowViewModel. Does this mean that there will be more and more logic here as the project gets bigger?
If so, should there be a static reference to the MainWindowViewModel that can be easily accessed by the other Viewmodels?
I'm a Unity programmer new to desktop development, sorry if my question made you feel stupid...

**Comment by thevortexcloud** [comment, marked as answer]
It really depends on how you structure your code and your presentation pattern. If you are going for a classical desktop approach you can just make it handle opening a bunch of sub windows each with their own viewmodels that are either self contained or return a result back to the main view model. If you are going for a SPA (which is required for some platforms such as mobile and web), it will typically just act as some sort of content presenter that hosts other view models and responds to page change requests. For simple apps you can also shove everything into the main view model. 

> Does this mean that there will be more and more logic here as the project gets bigger?

The more maintainable and flexible approach is to break as much as your code up into services which can be passed around. EG have a service for fetching data from a database, one for reading a file in a specific format, one for saving/reading user configuration, etc. Each service should be well encapsulated and not _directly_ depend on other services. They can indirectly depend on them though (EG an interface that a service implements).

You don't have to do things this way though. But if you do you can use dependency injection to make it easier.


> If so, should there be a static reference to the MainWindowViewModel that can be easily accessed by the other Viewmodels?

Typically you would not do that. Each view model should have minimal coupling to other view models. This makes reusing them easier. EG the main view model can instantiate child view models and pass in any dependent services.

> I'm a Unity programmer new to desktop development, sorry if my question made you feel stupid..

Not a stupid question at all. Unity does things fairly strangely compared to traditional C# (to the point I always have to warn normal C# devs about it). ECS is cool but almost unheard of in desktop development.

**Reply by coinsoundsbetter** [reply]
Thank you for your answer, which has improved my understanding of code organization in desktop development. In game development, we often need to access a singleton or static reference to accomplish many requirements. Therefore, when I do ViewModel development for subviews, I subconsciously want to give me direct access to the main window's reference by setting up singleton or static references so that I can pass some information to the MainWindowViewModel，Now I realize that was wrong.
Can you recommend some examples of open source that fit the desktop development style? For the official Avalonia examples, each one is undeniably lean and graceful. But I still have questions about the overall organizational structure of the project, such as: are there other design patterns for desktop development besides MVVM, how should such design be organized, etc... I read the Avalonia documentation, but my lack of knowledge in this area often left me at a loss as to where to start when organizing logic. - I might have been able to hard-write a less complex interface, but my code writing must have been a mess at this point.

**Reply by thevortexcloud** [reply]
> we often need to access a singleton or static reference to accomplish many requirements

God classes are kind of a requirement with how a lot of games are designed. In traditional C# development, people will look at you funny for even suggesting them. You can still use statics if you want to though. It just means your code is coupled to the static, which can make reusing it harder/writing unit tests impossible. I personally try my hardest to avoid them as a result and just use DI to achieve the same thing. But it can't always be avoided.

> there other design patterns for desktop development besides MVVM

Yes, several. MVVM just works best with Avalonia and most other XAML frameworks since it's what it is designed to be used with. You can also just ignore MVVM/any pattern entirely and do all your logic in the code behind of the view (which is how winforms works). You just wont have easy access to things like bindings if you do that. 

But in general (including things outside the XAML family) things like MVC, MVP and MVU are also common. For people who don't care about how their code works, the "throw code at the UI until it works" is also fairly popular. Their apps also coincidently tend to be a nightmare to maintain.

> Can you recommend some examples of open source that fit the desktop development style?

A complete and somewhat complicated example is the music store app tutorial. It guides you through all the basics you need to know of MVVM, ReactiveUI and Avalonia.

https://docs.avaloniaui.net/docs/tutorials/music-store-app/

The source code is here:
https://github.com/AvaloniaUI/Avalonia.MusicStore

It can be fairly confusing though if you have never touched MVVM before as it tries very hard to strictly follow the pattern.

If you want an actual practical example, the github credential manager app is actually written in C# and uses Avalonia:

https://github.com/git-ecosystem/git-credential-manager

But it's got a fairly weird structure, but it does follow MVVM to some extent  as far as I can tell!

> but my lack of knowledge in this area often left me at a loss as to where to start when organizing logic

If it helps you can also look at WPF examples. WPF and Avalonia are _very_ similar. To the point you can sometimes copy and paste code between them with minimal changes and it will probably work. I sometimes even just use the WPF documentation when Avalonia's is missing something. It does not always work, but it's still pretty useful to know that it's an option.

**Reply by coinsoundsbetter** [reply]
Yes, at first using MVVM did make me feel "limited", I believe that with continuous practice, I will break away from the original thinking, feel its robustness, thank you for your advice, thank you! Wish you every success！

**Reply by thevortexcloud** [reply]
You are welcome. MVVM also confused me a lot the first time I used it. But it's actually pretty nice if you follow it properly. It makes it very easy to port your entire app to a different UI framework if you need to. Also get into the habit of designing your code to be unit testable (DI and the service pattern I mentioned earlier helps with this). Even if you don't write unit tests it can still make your code far more flexible.
